### PR TITLE
Convert shares

### PR DESCRIPTION
**Pull request by benjaminsavage**
Next diff in the stack. This one uses the randomly generated pairs of secret shares of "r" in order to convert a u64 into an array of 64 secret sharings of the bits of that number.

Screenshot 2022-07-18 at 23 32 39

### CODE REVIEW

**Review by martinthomson — COMMENTED** [review]
I'm not sure about the arrangement of code into steps, but I guess without async/await, that is what you get.  A few minor tips though

File: src/modulus_convert.rs [hunk 1]
```diff
@@ -0,0 +1,1018 @@
...
+    <Self as ReplicatedSecretSharing>::Ring,
+);
+
+impl ReplicatedSecretSharing for ReplicatedFp31SecretSharing {
+    type Ring = Fp31;
+}
+
+impl ReplicatedFp31SecretSharing {
+    #[must_use]
+    pub fn construct(a: Fp31, b: Fp31) -> Self {
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply src/modulus_convert.rs:80]: I did not have an Into. In this PR (https://github.com/martinthomson/raw-ipa/pull/47/files) just did:

    pub fn new(a: T, b: T) -> Self {
        Self(a, b)
    }

I'm happy to add the "impl Into<T>" in place of just "T" if you think that's better.

File: src/modulus_convert.rs [hunk 2]
```diff
@@ -0,0 +1,1018 @@
...
+        if right_sent_d {
+            right_share += s1;
+        }
+        (Self::construct(left_share, right_share), d)
+    }
+
+    #[must_use]
+    pub fn mult_step2(
+        incomplete_shares: ReplicatedFp31SecretSharing,
+        d_value_received: Fp31,
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply src/modulus_convert.rs:117]: What do you think about the approach used here: https://github.com/martinthomson/raw-ipa/pull/55

We lost the optimization (for now) but that's just a temporary situation.

File: src/modulus_convert.rs [hunk 3]
```diff
@@ -0,0 +1,1018 @@
...
+            HelperIdentity::H3 => {
+                ReplicatedFp31SecretSharing::construct(Fp31::from(0_u8), Fp31::from(1_u8))
+            }
+        }
+    }
+
+    #[must_use]
+    pub fn convert_shares_step_1(
+        input: u64,
+        r_shares: &[(ReplicatedBinarySecretSharing, ReplicatedFp31SecretSharing)],
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply src/modulus_convert.rs:266]: Good catch.

**Comment by benjaminsavage** [comment]
I'm sorry I didn't close this old PR earlier! I already made most of these changes in https://github.com/martinthomson/raw-ipa/pull/47/files. Sorry for forgetting to close this!

File: src/modulus_convert.rs [hunk 4]
```diff
@@ -0,0 +1,1018 @@
...
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct ReplicatedFp31SecretSharing(
+    <Self as ReplicatedSecretSharing>::Ring,
+    <Self as ReplicatedSecretSharing>::Ring,
+);
+
+impl ReplicatedSecretSharing for ReplicatedFp31SecretSharing {
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply src/modulus_convert.rs:74]: Yup =). Did so here: https://github.com/martinthomson/raw-ipa/pull/47/files

File: src/modulus_convert.rs [hunk 5]
```diff
@@ -0,0 +1,1018 @@
...
+
+impl Sub for ReplicatedBinarySecretSharing {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        Self(self.0 - rhs.0, self.1 - rhs.1)
+    }
+}
+
+impl RandomShareGenerationHelper {
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply src/modulus_convert.rs:191]: Also implemented here: https://github.com/martinthomson/raw-ipa/pull/47/files

File: src/modulus_convert.rs [hunk 6]
```diff
@@ -0,0 +1,1018 @@
...
+
+    // Yeah, yeah... too many arguments... but it's just a test
+    #[allow(clippy::too_many_arguments)]
+    fn multiply_secret_shares(
+        h1: &RandomShareGenerationHelper,
+        h2: &RandomShareGenerationHelper,
+        h3: &RandomShareGenerationHelper,
+        a1: ReplicatedFp31SecretSharing,
+        a2: ReplicatedFp31SecretSharing,
+        a3: ReplicatedFp31SecretSharing,
```
[the remark this answers is not in the document]
Comment (benjaminsavage) [reply src/modulus_convert.rs:462]: I did so here: https://github.com/martinthomson/raw-ipa/pull/47/files

**Pull request closed without merging**